Panic: Runtime Error after Upgrade to Vault 1.1.1

We have upgraded our vault server from 1.0.0-beta2 to vault 1.1.1 including automated unseal using gcp. After the server upgrade, the new server ran for 3 days, until we noticed that it was not able to issue any new gcp credentials.

When trying to force-revoke any leases related to gcp, the vault server just timed out. Then we tried to increase the log level on the vault instances. After a restart of the master node, the vault server crashed with the following error message.

```
panic: runtime error: invalid memory address or nil pointer dereference[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x160f858]

goroutine 944 [running]:
github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin.(*backend).credentials.func1(0x25cfa60, 0xc000d364b0, 0x2b58a51, 0xb)
	/gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin/backend.go:148 +0xd8
github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin/cache.(*Cache).Fetch(0xc0007deda0, 0x2b58a51, 0xb, 0x1a3185c5000, 0xc0007df000, 0x0, 0xc000bdd560, 0x716de7, 0xc000d36e10)
	/gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin/cache/cache.go:58 +0x11f
github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin.(*backend).credentials(0xc000d364e0, 0x2f1ae60, 0xc0000114c0, 0xa0, 0x28061a0, 0x1)
	/gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin/backend.go:137 +0xab
github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin.(*backend).HTTPClient(0xc000d364e0, 0x2f1ae60, 0xc0000114c0, 0x40000c000eec3f0, 0x3, 0x0)
	/gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin/backend.go:114 +0x43
github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin.(*backend).IAMClient(0xc000d364e0, 0x2f1ae60, 0xc0000114c0, 0x8, 0xc000e915e8, 0xc000bdd601)
	/gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin/backend.go:90 +0x43
github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin.(*backend).secretKeyRevoke(0xc000d364e0, 0x2f1a120, 0xc000d36450, 0xc000ce6b40, 0xc000cf76b0, 0x2758580, 0x2b59301, 0xc000cf76b0)
	/gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin/secrets_service_account_key.go:157 +0xb8
github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin.(*backend).secretKeyRevoke-fm(0x2f1a120, 0xc000d36450, 0xc000ce6b40, 0xc000cf76b0, 0x2, 0xc000d36960, 0xc000cf7470)
	/gopath/src/github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-secrets-gcp/plugin/secrets_service_account_key.go:38 +0x52
github.com/hashicorp/vault/logical/framework.(*Secret).HandleRevoke(0xc000d36960, 0x2f1a120, 0xc000d36450, 0xc000ce6b40, 0xc000e91618, 0xc0004bd601, 0x1)
	/gopath/src/github.com/hashicorp/vault/logical/framework/secret.go:87 +0x94
github.com/hashicorp/vault/logical/framework.(*Backend).handleRevokeRenew(0xc0004bd5f0, 0x2f1a120, 0xc000d36450, 0xc000ce6b40, 0x0, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/logical/framework/backend.go:415 +0x1da
github.com/hashicorp/vault/logical/framework.(*Backend).HandleRequest(0xc0004bd5f0, 0x2f1a120, 0xc000d36450, 0xc000ce6b40, 0x0, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/logical/framework/backend.go:170 +0x845
github.com/hashicorp/vault/builtin/plugin.(*PluginBackend).HandleRequest(0xc0007d7220, 0x2f1a120, 0xc000d36450, 0xc000ce6b40, 0x0, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/builtin/plugin/backend.go:161 +0xc7
github.com/hashicorp/vault/vault.(*Router).routeCommon(0xc0001e87c0, 0x2f1a120, 0xc000d36450, 0xc000ce6b40, 0x40c100, 0x0, 0x2af0000, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/vault/router.go:665 +0x926
github.com/hashicorp/vault/vault.(*Router).Route(0xc0001e87c0, 0x2f1a120, 0xc000d36450, 0xc000ce6b40, 0xc000d36450, 0xc000bddd48, 0x1f891b1)
	/gopath/src/github.com/hashicorp/vault/vault/router.go:465 +0x4e
github.com/hashicorp/vault/vault.(*ExpirationManager).revokeEntry(0xc000735d40, 0x2f1a120, 0xc000776360, 0xc0012c9c20, 0x40, 0xc0012c9c20)
	/gopath/src/github.com/hashicorp/vault/vault/expiration.go:1363 +0x284
github.com/hashicorp/vault/vault.(*ExpirationManager).revokeCommon(0xc000735d40, 0x2f1a120, 0xc000776360, 0xc0007e8d00, 0x40, 0xbf27ef231cbc0000, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/vault/expiration.go:626 +0x4b3
github.com/hashicorp/vault/vault.(*ExpirationManager).Revoke(0xc000735d40, 0x2f1a120, 0xc000776360, 0xc0007e8d00, 0x40, 0x0, 0x0)
	/gopath/src/github.com/hashicorp/vault/vault/expiration.go:573 +0x11c
github.com/hashicorp/vault/vault.expireLeaseStrategyRevoke(0x2f1a060, 0xc000832480, 0xc000735d40, 0xc0008f0be0)
	/gopath/src/github.com/hashicorp/vault/vault/expiration.go:122 +0x2c4
github.com/hashicorp/vault/vault.(*ExpirationManager).updatePendingInternal.func1()
	/gopath/src/github.com/hashicorp/vault/vault/expiration.go:1323 +0x52
created by time.goFunc
	/goroot/src/time/sleep.go:172 +0x44
```

We have been using GCP Roles defined using JSON before (see #27) probably that might be related to the issue. Here is one of our (terraform) configurations for creating a GCP role.

```
resource "vault_generic_secret" "gcp_roleset_application" {
  path         = "${var.gcp_secret_backend_path}/roleset/application"
  disable_read = true

  data_json = <<EOT
{
  "project":   "${var.project_id}",
  "secret_type": "service_account_key",
  "bindings": "resource \"projects/${var.project_id}/topics/${var.pubsub_topic_1_name}\" { roles = [\"roles/pubsub.publisher\"] } resource \"buckets/${var.storage_bucket_1_name}\" { roles = [\"roles/storage.objectViewer\"] } resource \"projects/${var.project_id}/topics/${var.pubsub_topic_name}\" { roles = [\"roles/pubsub.publisher\"] } resource \"buckets/${var.storage_bucket_2_name}\" { roles = [\"roles/storage.objectViewer\"] } resource \"projects/${var.project_id}/topics/${var.pubsub_topic_2_name}\" { roles = [\"roles/pubsub.publisher\"] }"
}
EOT
}
```

We ended up downgrading the vault server again.

--- a/plugin/backend.go
+++ b/plugin/backend.go
@@ -139,13 +139,15 @@ func (b *backend) credentials(s logical.Storage) (*google.Credentials, error) {
 
 		ctx := context.Background()
 
-		config, err := getConfig(ctx, s)
+		cfg, err := getConfig(ctx, s)
 		if err != nil {
 			return nil, err
 		}
-
+		if cfg == nil {
+			cfg = &config{}
+		}
 		// Get creds from the config
-		credBytes := []byte(config.CredentialsRaw)
+		credBytes := []byte(cfg.CredentialsRaw)
 
 		// If credentials were provided, use those. Otherwise fall back to the
 		// default application credentials.

--- a/plugin/util/parse_bindings.go
+++ b/plugin/util/parse_bindings.go
@@ -107,7 +107,7 @@ func parseResourceObject(item *ast.ObjectItem, bindings map[string]StringSet) er
 
 	resourceItemList := item.Val.(*ast.ObjectType).List
 	if resourceItemList == nil {
-		return fmt.Errorf("invalid empty roles list for item", item.Assign.Line)
+		return fmt.Errorf("invalid empty roles list for item (line %d)", item.Assign.Line)
 	}
 
 	var merr *multierror.Error
@@ -137,7 +137,7 @@ func parseRolesObject(rolesObj *ast.ObjectItem, parsedRoles StringSet) error {
 	}
 	roleList, ok := rolesObj.Val.(*ast.ListType)
 	if !ok {
-		return fmt.Errorf("parsing error, expected list of roles for key %q")
+		return fmt.Errorf("parsing error, expected list of roles for key 'roles'")
 	}
 	var merr *multierror.Error
 	for _, singleRoleObj := range roleList.List {

